app/src/main/java/org/wikipedia/page/tabs/TabActivity.kt

```diff
@@ -328,35 +195,83 @@ class TabActivity : BaseActivity() {
-            clearFirstTabBitmap()
-            try {
-                if (view.isLaidOut) {
-                    FIRST_TAB_BITMAP = view.drawToBitmap(Bitmap.Config.RGB_565)
-                    FIRST_TAB_BITMAP_TITLE = title
-                }
-            } catch (e: OutOfMemoryError) {
-                // don't worry about it
+    private open inner class TabViewHolder(itemView: View) : RecyclerView.ViewHolder(itemView), View.OnClickListener, SwipeableTabTouchHelperCallback.Callback {
+        open fun bindItem(tab: Tab, position: Int) {
+            itemView.findViewById<TextView>(R.id.tabArticleTitle).text = StringUtil.fromHtml(tab.backStackPositionTitle?.displayText.orEmpty())
+            itemView.findViewById<TextView>(R.id.tabArticleDescription).text = StringUtil.fromHtml(tab.backStackPositionTitle?.description.orEmpty())
+            itemView.findViewById<View>(R.id.tabContainer).setOnClickListener(this)
+            itemView.findViewById<View>(R.id.tabCloseButton).setOnClickListener(this)
+            itemView.findViewById<WikiCardView>(R.id.tabCardView).run {
+                strokeWidth = DimenUtil.roundedDpToPx(if (position == 0) 1.5f else 1f)
+                strokeColor = ResourceUtil.getThemedColor(context, if (position == 0) R.attr.progressive_color else R.attr.border_color)
             }
         }
 
-        private fun clearFirstTabBitmap() {
-            FIRST_TAB_BITMAP_TITLE = ""
-            FIRST_TAB_BITMAP?.run {
-                if (!isRecycled) {
-                    recycle()
+        override fun onClick(v: View) {
+            val adapterPosition = bindingAdapterPosition
+            val index = adapterPositionToTabIndex(adapterPosition)
+            if (index < 0 || index >= app.tabList.size) {
+                return
+            }
+            if (v.id == R.id.tabContainer) {
+                if (index < app.tabList.size - 1) {
+                    val tab = app.tabList.removeAt(index)
+                    app.tabList.add(tab)
                 }
-                FIRST_TAB_BITMAP = null
+                cancelled = false
+                if (launchedFromPageActivity) {
+                    setResult(RESULT_LOAD_FROM_BACKSTACK)
+                } else {
+                    startActivity(PageActivity.newIntent(this@TabActivity))
+                }
+                finish()
+            } else if (v.id == R.id.tabCloseButton) {
+                doCloseTab()
             }
         }
 
+        override fun onSwipe() {
+            doCloseTab()
+        }
+
+        override fun isSwipeable(): Boolean {
+            return true
+        }
+
+        private fun doCloseTab() {
+            val adapterPosition = bindingAdapterPosition
+            val index = adapterPositionToTabIndex(adapterPosition)
+            val appTab = app.tabList.removeAt(index)
+            binding.tabCountsView.updateTabCount(false)
+            bindingAdapter?.notifyItemRemoved(adapterPosition)
```

Comment:
We user removes the first tab from the list (the one with a blue border), the second tab moves to the first position, but does not get the border set.

Reply:
Done!